### ISSUE TITLE
ZIndex isn't respected properly on ios

### ISSUE THREAD

**Issue by Hari-07** [flutter/flutter#155897]
### What package does this bug report belong to?

google_maps_flutter

### What target platforms are you seeing this bug on?

iOS

### Have you already upgraded your packages?

Yes

### Dependency versions

<details><summary>pubspec.lock</summary>

```lock
[Paste file content here]
```

</details>

### Steps to reproduce

1. Copy paste the code and open it

### Expected results

The marker with the higher zIndex should be drawn above the other

### Actual results

The marker thats second in the array is drawn above

### Code sample

<details open><summary>Code sample</summary>

```dart
class MapZIndexExample extends StatelessWidget {
  const MapZIndexExample({super.key});

  @override
  Widget build(BuildContext context) {
    return GoogleMap(
      compassEnabled: false,
      initialCameraPosition: const CameraPosition(
        target: LatLng(37.334542, -122.009325),
        zoom: 12,
      ),
      myLocationButtonEnabled: false,
      zoomControlsEnabled: false,
      minMaxZoomPreference: const MinMaxZoomPreference(3, 20),
      markers: {
        const Marker(
          markerId: MarkerId('1'),
          position: LatLng(37.334542, -122.014325),
          zIndex: 3.7,
        ),
        const Marker(
          markerId: MarkerId('2'),
          position: LatLng(37.334542, -122.009325),
          zIndex: 3.5,
        ),
      },
    );
  }
}
```

</details>

### Screenshots or Videos

<details open>
<summary>Screenshots / Video demonstration</summary>

<img src="https://github.com/user-attachments/assets/ee8565d2-7b84-404e-87bc-96fed37f5db4" height="600">

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

**Comment by Hari-07** [comment]
If you look at [google maps flutter ios package](https://github.com/flutter/packages/blob/27c98537ccac576cfb79e9407a5afc0c36667f7d/packages/google_maps_flutter/google_maps_flutter_ios/ios/Classes/GoogleMapMarkerController.m#L150-L152) , we can see it seems like its truncating the double into an int. That's very confusing for a user of the package.

```objc
  NSNumber *zIndex = FGMGetValueOrNilFromDict(data, @"zIndex");
  if (zIndex) {
    [self setZIndex:[zIndex intValue]];
  }
```

**Comment by stuartmorgan** [comment]
> it looks to me like its truncating the double into an int

That's because `zIndex` is an integer property in the iOS Google Maps SDK.

It looks like this was a mistake in how `zIndex` was added to the plugin; the right approach would have been to smooth out this platform difference in our abstraction by making our `zIndex` an integer as well, since in practice that would have allowed for consistent behavior across all the implementations.

Fixing it thay way now would be a breaking version change. We can consider that for the long term, but in the short term we should consider either:
1. Documenting this limitation (not great, but certainly better than the current situation)
2. Internally scaling the zIndex on iOS to try to patch over this. E.g., if we multiply the incoming zIndex by 1000 on iOS (and reverse that when reading back), differences out to three digits would be respected. We'd have to add a safety check that we wouldn't introduce int overflow, but I can't imagine that any would, in practice, both use numbers big enough for that to happen and also use non-integer values in the first place, so that would probably work out.

We'd probably need to document something even if we do 2, such as that only a few decimal places are honored, which may mean it's not worthwhile to do 2 in the first place. iOS folks, what do you think?

**Comment by Hari-07** [comment]
Mutliple google maps related params already have varying or partial functionality depending on ios/android (Some params dont have any effect on one platform). So documenting this would be more effective than it generally is to begin with, since its common in the package

Another solution I thought off, but not sure if it would be performant enough is that in the ios layer we could just order the markers by the passed in zIndex and then give it zIndex in serial order. This would maintain the order as expected via the double zIndex but not use the actual value the user passed in

**Comment by Leoporno** [comment]
****

**Comment by stuartmorgan** [comment]
Reid points out that maybe the least worst option is to add a new int version, and deprecate the old one, which is non-breaking (at the cost of a less clean API surface).

https://github.com/flutter/flutter/issues/155526 may also give us a better longer-term avenue to fix this, as we should be able to avoid this problem in the new marker (unless it ends up as a subclass).

**Comment by Hari-07** [comment]
Advanced markers probably wont be out for a while right? This should probably be fixed sooner than that. And yeah when doing that it should definitely be an int param (Should this be left as a comment on that issue to make sure its not forgotten?)

**Comment by jmagman** [comment]
> This should probably be fixed sooner than that.

@stuartmorgan do you think adding a new int version and deprecating the old double should be blocking for some reason on the markers fix?  Or can we address it sooner?

If so, @Hari-07 would you be willing to tackle it? I recall your excellent polygon work 🙂 
https://github.com/flutter/packages/pull/5757

**Comment by Hari-07** [comment]
@jmagman Sure, I was actually gonna tackle it, but seemed like there wasn't a consensus with which direction to go with.

I do think its reasonable that we deprecate this field, and add another int field for the interim while advanced markers isn't out yet. Just wasn't sure if @stuartmorgan was onboard with that or not so was waiting for an update lol

**Comment by stuartmorgan** [comment]
> @stuartmorgan do you think adding a new int version and deprecating the old double should be blocking for some reason on the markers fix? Or can we address it sooner?

I don't think we should block on advanced markers; if they end up as a subclass we'll need this regardless, and if they don't then we can likely eventually deprecate the old markers so it will be less problematic that we have a kind of weird API surface on them (with the double and int versions).

**Comment by jmagman** [comment]
@Hari-07 sounds like a PR to deprecate and replace an int field would be welcome then!
